documentation/proposals/Proposal - Better Strings.md

```diff
@@ -0,0 +1,102 @@
+# Summary
+A rich, more well-defined native string API.
+
+- Written against Silk.NET **1.0.2**.
+- Slated for Silk.NET **2.0.0**.
+
+# Contributors
+- Dylan P, Ultz Limited
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+Currently, the representation of native strings in Silk.NET (and C# as a whole) is an utter mess.
+At the moment we're using `char*` everywhere which isn't technically correct, as a number of users
+have pointed out. This discussion aims to create a richer set of native string APIs to combine
+technical correctness, ease of marshalling, and ease of use.
+
+This is in the form of Char8, Char16, and Char32 which represent the various lengths of characters
+in various encodings. These CharXXs can be combined into a StringXX. All types can be implicitly
+converted to and from their existing C#/.NET types, so that the user doesn't have to deal with much
+
+# Proposed API
+Proposed Silk.NET.Core additions:
+```cs
+public struct Char8
+{
+    private byte _value; // 8-bit character
+    public Char8(byte v) => _value = v;
+    public static implicit operator Char8(char c);
+    public static implicit operator Char8(byte c);
+    // and vice versa
+    public char ToChar();
+    public byte ToByte();
+    public uint ToUInt32();
+}
+
+public struct Char16
+{
+    private char _value; // 16-bit character
+    public Char16(char v) => _value = v;
+    // insert appropriate operators and methods here as seen in Char8
+}
+
+public struct Char32
+{
+    private uint _value; // 32-bit character
+    public Char32(char v0, char v1) => _value = ((uint)v0 << 16) + (uint)v1;
```

Comment:
This isn't a correct conversion as it doesn't account for surrogates.

Reply:
Basically UTF-16 is either a standalone 16-bit character or it is two surrogates
So its either a zero extension or a different algorithm that combines the surrogate halves (with appropriate checks that v0 is a high surrogate and v1 is a low).

Reply:
I would suggest using `char.ConvertToUtf32()` here, assuming that uses the correct algorithm.

Reply:
Yes it does: https://source.dot.net/#System.Private.CoreLib/Char.cs,890

Would just need to cast from `int` to `uint` for the result.